BUG: fix memory leak in JSON datetime serialization

Alvaro-Kothe opened the pull request:
- [x] closes #62204 (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Alvaro-Kothe committed:
fix: memory leak in JSON datetime serialization

swt2c reviewed (commented):
I think you might also need to add a `GET_TC(tc)->freeCStr = 1;` in `Object_getBigNumStringValue`

Alvaro-Kothe committed:
fix: flag to free `cStr` in `Object_getBigNumStringValue`

Alvaro-Kothe commented:
Thanks for pointing that out. Just fixed it.

swt2c commented on pandas/_libs/src/vendored/ujson/python/objToJSON.c:
```diff
@@ -107,6 +107,7 @@ typedef struct __TypeContext {
   JSINT64 longValue;
 
   const char *cStr;
+  int freeCStr;
```
Any suggestions on how to do that?  Revert the changes in https://github.com/pandas-dev/pandas/commit/fb6c4e33c45938d7675d4c9a132324cd08df2f3c to make some of the uses of `cStr` stack allocated?  It doesn't seem possible to make all uses of `cStr` stack allocated, at least at first glance.

Alvaro-Kothe committed:
fix: use dynamic allocated memory to store strings

Revert "fix: flag to free `cStr` in `Object_getBigNumStringValue`"

Revert "fix: memory leak in JSON datetime serialization"

fix: fix memory leak in `PyDateTimeToIsoCallback`

fix: ensure successful malloc.

WillAyd commented on pandas/_libs/src/vendored/ujson/python/objToJSON.c:
```diff
@@ -1880,6 +1905,7 @@ static void Object_endTypeContext(JSOBJ Py_UNUSED(obj), JSONTypeContext *tc) {
     GET_TC(tc)->rowLabels = NULL;
     NpyArr_freeLabels(GET_TC(tc)->columnLabels, GET_TC(tc)->columnLabelsLen);
     GET_TC(tc)->columnLabels = NULL;
+    PyObject_Free((void *)GET_TC(tc)->cStr);
```
I'm not too sure about this approach - generally the CPython API only wants you to call this function when you've allocated with the corresponding function(s)

https://docs.python.org/3/c-api/memory.html#c.PyObject_Free

> Frees the memory block pointed to by p, which must have been returned by a previous call to [PyObject_Malloc()](https://docs.python.org/3/c-api/memory.html#c.PyObject_Malloc), [PyObject_Realloc()](https://docs.python.org/3/c-api/memory.html#c.PyObject_Realloc) or [PyObject_Calloc()](https://docs.python.org/3/c-api/memory.html#c.PyObject_Calloc). Otherwise, or if PyObject_Free(p) has been called before, undefined behavior occurs.

Alvaro-Kothe committed:
fix: replace PyObject_Malloc with PyMem_Malloc

WillAyd reviewed (changes requested):
Are there any object types where it is possible to go through the destructor without initializing the `cStr` member? Looking at say the dict iteration code it seems like that may be happening?

Its a bit confusing here that the type context begin functions have an implicit requirement to allocate the `cStr` member. I wonder if it wouldn't be cleaner to allocate that up front, and leave it to individual functions to free or replace memory as required

Alvaro-Kothe committed:
fix: drop `const` qualifier in `cStr`

Alvaro-Kothe committed:
fix: remove void pointer case in PyMem_Free

Alvaro-Kothe commented on pandas/_libs/src/vendored/ujson/python/objToJSON.c:
```diff
@@ -1880,6 +1905,7 @@ static void Object_endTypeContext(JSOBJ Py_UNUSED(obj), JSONTypeContext *tc) {
     GET_TC(tc)->rowLabels = NULL;
     NpyArr_freeLabels(GET_TC(tc)->columnLabels, GET_TC(tc)->columnLabelsLen);
     GET_TC(tc)->columnLabels = NULL;
+    PyObject_Free((void *)GET_TC(tc)->cStr);
```
It was because of the const qualifier in `cStr`, but since I removed this qualifier, I am no longer casting.

Alvaro-Kothe committed:
fix: create const for default `cStr` size

Alvaro-Kothe commented on pandas/_libs/src/vendored/ujson/python/objToJSON.c:
```diff
@@ -1007,16 +1008,24 @@ static const char *List_iterGetName(JSOBJ Py_UNUSED(obj),
 //=============================================================================
 static void Index_iterBegin(JSOBJ Py_UNUSED(obj), JSONTypeContext *tc) {
   GET_TC(tc)->index = 0;
+  GET_TC(tc)->cStr = PyObject_Malloc(20);
```
I created the variable `CSTR_SIZE`

Alvaro-Kothe committed:
try to fix ci error

WillAyd commented on pandas/_libs/src/datetime/date_conversions.c:
```diff
@@ -49,7 +49,7 @@ char *int64ToIso(int64_t value, NPY_DATETIMEUNIT valueUnit,
 
   *len = (size_t)get_datetime_iso_8601_strlen(0, base);
-  char *result = PyObject_Malloc(*len);
+  char *result = PyMem_Malloc(*len);
```
Since the focus of the PR is to fix the JSON modules, let's leave the others untouched for now. These may be viable fixes, but they can always be done in a follow up

WillAyd replied:
Ah nevermind - I guess you are doing this because the JSON module makes use of this function

Alvaro-Kothe replied:
Exacly, I made this change because of the assignment of `cStr` in `NpyDateTimeToIsoCallback`.

 I also made changes in other functions to allocate with `PyMem_Malloc` in `int64ToIsoDuration` and `PyDateTimeToIso`, but I think that the CI errors are because of this changes. Is it correct to use `PyMem_Malloc` in them or should I change back to `PyObject_Malloc`?

WillAyd commented:
@Alvaro-Kothe the address and undefined behavior violations can be pretty tricky to debug. If you aren't aware, you can turn on sanitizers locally to better understand what is happening (assuming your compiler supports them):

```
python -m pip install -e . -Csetup-args="-Db_sanitize=address,undefined" -Csetupargs="-Dbuildtype=debug"
```

Alvaro-Kothe committed:
hack: go back to use `PyObject_*` because of CI errors

WillAyd reviewed (commented):
Cool nice work

Alvaro-Kothe commented on pandas/_libs/src/vendored/ujson/python/objToJSON.c:
```diff
@@ -1880,6 +1907,7 @@ static void Object_endTypeContext(JSOBJ Py_UNUSED(obj), JSONTypeContext *tc) {
     GET_TC(tc)->rowLabels = NULL;
     NpyArr_freeLabels(GET_TC(tc)->columnLabels, GET_TC(tc)->columnLabelsLen);
     GET_TC(tc)->columnLabels = NULL;
+    PyObject_Free(GET_TC(tc)->cStr);
```
I checked these paths where `cStr` is assigned to a dynamic memory allocated with `PyObject_Malloc`:
- `NpyDateTimeToIsoCallback -> int64ToIso`
- `NpyTimeDeltaToIsoCallback -> int64ToIsoDuration`
- `PyDateTimeToIsoCallback -> PyDateTimeToIso`
- `Index_iterBegin`
- `Series_iterBegin`
- `DataFrame_iterBegin`
- `Object_getBigNumStringValue`

There is no `*_Free` call for `cStr` outside of the `Object_endTypeContext` function to cause an undefined behaviour. 

Other cases seems like `cStr` isn't assigned and end up being `NULL`, making the `PyObject_Free` a [no-op](https://docs.python.org/3/c-api/memory.html#c.PyObject_Free).
